### PR TITLE
Add JS helper for registering theme modules

### LINKED ISSUE #49: Provide sane styling helpers for JS users
Right now, our recommendation for P3 users is that they wrap their CSS in HTML which is wrapped in JS. This double wrapping is really ridiculous, even though I understand why it's necessary.

I don't know if there's very much we can do for the bare bones case, but we can at least offer something significantly better for those who use a build system that can emulate the upcoming CSS Modules specification (e.g. webpack with `css-loader`).

Usage could look like this (names heavily pending):
```
import * as helpers from "@vaadin/vaadin-themable-mixin/helpers";
import globalStyles from "./globalStyles.css";
import gridStyles from "./gridStyles.css";

helpers.exportAsCustomStyles(globalStyles);
helpers.exportAsThemeModule("vaadin-grid", gridStyles);
```

### PR DESCRIPTION
**Pull request by tomivirkki**
Fixes #49

### CODE REVIEW

**Comment by jouni** [comment]
More more thought that just came up: could we avoid having to specify a unique ID for a reusable style sheet (aka style module), and instead use the direct DOM reference to the element?

Something like this:

```javascript
import { createStyle } from '@vaadin/vaadin-themable-mixin/styler.js';

let reusable = Vaadin.createStyle(`...`);

Vaadin.createStyle(`...`).include(reusable).registerForComponent('my-element');
```

Do we still need to have a “terminal command” for both cases (a reusable style sheet and a component style sheet) as we are always creating a `<dom-module>`, and it doesn’t really matter if the `theme-for` attribute is added after the element is already attached to the DOM?

The method chaining would not work then, though – unless we add those to the `<dom-module>` elements.

**Comment by tomivirkki** [comment]
With the current state of the PR, you can already consider the `reusable` object as shared and register that directly for multiple components:
```js
reusable.registerForComponent('my-element-one');
// Elsewhere
reusable.registerForComponent('my-element-two');
```
The underlying implementation should still be optimized not to create a new `<dom-module>` instance for every registration.

That said, is there even need for the `include()` API...? Would just `createStyle()` and `registerForComponent()` cover all the use cases?

**Comment by tomivirkki** [comment]
Hmm...what if the style rules were shared as plain string references...

```js
import { registerStyles } from '@vaadin/vaadin-themable-mixin/registerStyles.js';
import { reusable } from './my-shared-styles.js'; // a string reference

registerStyles('my-element', reusable);
```

**Comment by Legioth** [comment]
We can indeed automatically generate an `id` and attach the `<dom-module>` instance dynamically the first time an instance is passed to `include()`. To support chaining, the instance returned from `createStyle` could still be a wrapper rather than the original `<dom-module>` instance.

I suspect that we still also have to support the case of explicitly attaching with a given id to support referencing some reusable styles from a `<custom-style>` in the application's index html document, i.e. `<custom-style><style include="lumo-typography"></style></custom-style>`. Alternatively, we're back at creating some way of also creating your own `<custom-style>` with the same inputs as for the other purposes.

I'm not sure I understand how the idea with `registerStyles('my-element', reusable);` would work in case you want to combine your own CSS string with styles included from a reusable module. We also need to take into account the need for giving an explicit `id` in the `theme-for` case so that the component's styles can be applied in the right order.

**Comment by tomivirkki** [comment]
> I'm not sure I understand how the idea with registerStyles('my-element', reusable); would work in case you want to combine your own CSS string with styles included from a reusable module. 

The idea is to abstract away the concept of Polymer's style modules in general. The language provides built-in module system so why not use that instead? For users who want to reuse styles from existing Polymer modules with the new API (and for some reason can't convert them to strings), we could temporarily provide a means to extract the styles from them. For example:


```js
import { registerStyles, extractLegacyStyles } from '@vaadin/vaadin-themable-mixin/registerStyles.js';
import { reusable } from './my-shared-styles.js'; // string

const cssStringFromModule = extractLegacyStyles('module-name');

const myCSSstring = `
  :host {
    color: blue;
  }
`;

registerStyles('my-element', reusable, myCSSstring, cssStringFromModule);
```

..but the recommended way would be to not use style modules at all.

> We also need to take into account the need for giving an explicit id in the theme-for case so that the component's styles can be applied in the right order.

Identifying modules by a string based id from a global registry is also residue from the legacy module sharing system. The include order based on that id is also thus a bit outdated. An explicit priority number or such would make more sense with the suggested API.

**Comment by Legioth** [comment]
So after this PR is merged, what would the application developer write in each of these cases:

1. Use `lumo-typography` and make all links with `class=special` green in the top-light DOM
2. Use `lumo-typography` and make all links with `class=special` green in the shadow DOM of `my-view`.
3. Use `lumo-typography` and make all links with `class=special` green in the shadow DOM of a third-party `special-button` component that uses `VaadinThemableMixin`?

**Comment by tomivirkki** [comment]
> Use lumo-typography and make all links with class=special green in the top-level light DOM

As we agreed earlier, we don't intend to provide helper APIs for creating `<custom-style>`s as it has nothing to do with themable-mixin. Even the original API proposal has no helpers for `<custom-style>`.

> Use lumo-typography and make all links with class=special green in the shadow DOM of my-view.

Assuming my-view extends themable-mixin:

```js
// With conversion of lumo-modules
import { lumoTyphography } from '@vaadin/vaadin-lumo-styles/typography';
// Without conversion of lumo-modules
const lumoTyphography = extractLegacyStyles('lumo-typography');

registerStyles('my-view', lumoTyphography , `
  a.special {
    color: green;
  }
`)
```

> Use lumo-typography and make all links with class=special green in the shadow DOM of a third-party special-button component that uses VaadinThemableMixin?

```js
registerStyles('special-button', lumoTyphography , `
  a.special {
    color: green;
  }
`)
```

I think it mostly comes down to how much we want to be forwards compatible with the API. If it's just a helper for the current techincal implementation then the original API is probably the way to go. But Polymer will get dropped on some timeframe, and at that point, it would end up being Vaadin's proprietary module system based on global string keys. Or we'd have to break the API again.

**Comment by Legioth** [comment]
`import { lumoTyphography } from '@vaadin/vaadin-lumo-styles/typography';` implies (but doesn't explicitly say so) that `@vaadin/vaadin-lumo-styles/typography` has no side effects such as attaching an `<dom-module id="lumo-typography">` element. Without that side effect, the currently recommended approach of including `<custom-style><style include="lumo-typography"></style></custom-style>` in your index.html would no longer work as-is.

**Comment by tomivirkki** [comment]
Discussed with @Legioth in person. Eventually, the theming API could look somewhat as follows:
```js
import { registerStyles } from '@vaadin/vaadin-themable-mixin';
import typography from '@vaadin/lumo/typography.css';

registerStyles('special-button', typography, `
  a.special {
    color: green;
  }
`);
```

For now, as we still need a way to support including the existing Polymer style modules, we can extend the API as follows:

```js
registerStyles('special-button', 'lumo-typography', `
  a.special {
    color: green;
  }
`);
```

...so that any argument string that doesn't have CSS rules is treated internally as a style module id. Existence of character "{" would be used to determine whether a string parameter is a set of CSS rules or a module id.

Documentation for the feature: https://github.com/vaadin/vaadin-themable-mixin/pull/56/files#diff-04c6e90faac2675aa89e2176d2eec7d8R161

File: README.md [hunk 1]
```diff
@@ -158,7 +158,46 @@ The `id` attribute of the theme module should be unique. You can also re-use an
...
+
+Use the `registerStyles(themeFor, ...styles)` function to register CSS styles to be included in a component's local scope. The `themeFor` parameter is of type string and is used to identify the component type the styles are applied to. It works the same as with style modules. The `...styles` accepts an arbitrary list of string type parameters, each of which contains CSS style rules to be included in the targeted component's local style scope.
+
+```js
+Vaadin.registerStyles('my-element', `
+  /* Styles which will be included in my-element local scope */
+`);
+```
+
+> Note: If you need to include styles defined as style modules with id, you can still pass the id as a paramter for `...styles` to get the module included. Consider this API deprecated though.
```
Comment (DiegoCardoso) [inline README.md]: Typo in `paramter`.

Comment (tomivirkki) [reply README.md]: Done.

**Comment by platosha** [comment]
I have a concern for accepting strings with module names and strings with style contents in the same argument of `registerStyles`. The drawbacks here are:
- Requires long textual explanation in docs about what the accepted string formats are.
- Hard to describe this in JSdoc and in static type systems (TypeScript, Java). Developers would likely have poor IDE suggestions because of that.
- Static analysis tools would not be able to extract the styles from the code.

I think, relying only on runtime discretion between string formats with different meaning is not good for a high-level API. It would be much nicer have more abstractions rather than more string formats. 

How about only accepting ``` styleModule`lumo-styles` ``` and ``` css`a.special {}` ``` literals?

**Comment by tomivirkki** [comment]
One drawback is that then we'd provide explicit API for using legacy modules which would later force us to break the API (by removing the `styleModule` tag) once Polymer dependencies are dropped. From a developer's point of view this might still be better than magic strings as he'd then notice right away why his code no longer works.

Would ```registerStyles('special-button', `a.special {}`, styleModule('my-old-styles'))``` make sense (`styleModule` for Polymer modules, strings for normal styles). Then later we'd remove `styleModule` function?

Any suggestion for https://github.com/vaadin/vaadin-themable-mixin/pull/56/files#diff-04c6e90faac2675aa89e2176d2eec7d8R191 ? Mixing strings and objects in `...styles` also kind of feels wrong. ```registerStyles(registerAsModule('my-style-module'), `a.special {}`, styleModule('my-old-styles'))```?

**Comment by platosha** [comment]
With `...styles` as a rest argument, it is technically allowed to have zero length:

```javascript
function registerStyles(themeFor, ...styles) {
}

registerStyles('x-foo'); // valid code, my IDE didn’t complain, huh?
```

Also, the rest argument is required to be the last. If it is desired to support an optional options object as the last argument, JSdoc or TypeScript would require declaring a mixed type for the `...rest` argument, so that it could contain zero or more items with either options or style contents.

A single explicit style contents argument would have several advantages here:
- IDEs will remind developers to pass it
- It is compatible with further optional arguments, e. g., optional `{moduleId: string}`
- Single argument does not limit to single style content string, concatenating them should be quite easy
- Could be an `StyleContent|Array<StyleContent>` to support passing many style contents at once, if needed.

**Comment by platosha** [comment]
`LitElement` has quite reasonable abstractions for styles, btw:
- The [`styles`](https://github.com/Polymer/lit-element/blob/4b0c88c26a1b64ab63edfe5deb705a3a7cb26a5f/src/lit-element.ts#L81-L85) static base class property does only support instances of `CSSResult` (single instance or array). The library **would not accept plain strings** as styles, this throws.
- `CSSResult` is not constructable, [its constructor throws](https://github.com/Polymer/lit-element/blob/4b0c88c26a1b64ab63edfe5deb705a3a7cb26a5f/src/lib/css-tag.ts#L24-L27).
- The only way to obtain a `CSSResult` is to use one of the supplied literals: `css` or `unsafeCSS`

Some advantages are:
- Requires to annotate the plain string with `css` in the code, giving context for IDEs, tools, readers.
- The `css` tagged string ensures that every string interpolation there is annotated too, e. g., ``` css`${'/* plain nested string */'}` ``` throws, ``` css`${ css`/* annotated nested string */` } ``` is ok
- The template string syntax makes concatenation easy. There is no much need for an array if you can concatenate your style strings easily.

**Comment by tomivirkki** [comment]
> ...How about only accepting styleModule`lumo-styles` and css`a.special {}` literals?"

The API tweaks that deal with Polymer style modules shouldn't be considered a part of the official API but something that you _can_ do (while discouraged). It might still be better not to provide dedicated tags etc for them but rather documentation on how to hack style modules in if really necessary.

> With ...styles as a rest argument, it is technically allowed to have zero length

We can show a warning if someone accidentally does this.

> Single argument does not limit to single style content string, concatenating them should be quite easy

Concatenating the string styles should rather be discouraged as the union of string styles X and Y would then need to be implemented as an inclusion of a new style module XY, instead of separate inclusions of X and Y which can be optimized.

**Comment by web-padawan** [comment]
I'm generally in favour of using `css` tagged literal, based on the following:

1. new API is supposed to fix the DX issues with CSS which Polymer 3 has, especially:
  -  syntax highlighting in VSCode using [es6-string-css](https://marketplace.visualstudio.com/items?itemName=bashmish.es6-string-css)
  -  ability to run Stylelint using [stylelint-processor-styled-components](https://github.com/styled-components/stylelint-processor-styled-components)
   - ability to format CSS code in tagged literals using Prettier
   - ability to minify CSS using [babel-plugin-template-html-minifier](https://github.com/cfware/babel-plugin-template-html-minifier)

Most of the above tools rely on the *tagged* literal presence. AFAIK only `es6-string.css` supports `/* css */` comments as well. So with `css` literal we will have better DX.

2. new API is purely for web components and we don't have to make it work with Java
  - no need to take care of how tagged literals would be created on the server

**Comment by tomivirkki** [comment]
The security issue mentioned by @platosha could not be refuted. Updated the PR to include the `css` tag and a separate `unsafeCSS` function

File: README.md [hunk 2]
```diff
@@ -158,7 +158,61 @@ The `id` attribute of the theme module should be unique. You can also re-use an
...
+```html
+<link rel="import" href="../register-styles.html">
+```
+
+Importing the helper (as an ES module)
+```js
+import { registerStyles, css } from '@vaadin/vaadin-themable-mixin/register-styles.js';
+```
+
+Use the `registerStyles(themeFor, styles)` function to register CSS styles to be included in a component's local scope. The `themeFor` parameter is of type string and is used to identify the component type the styles are applied to. It works the same as with style modules. The `styles` accepts a an object or an array of objects built as template literals tagged with `css`, containing CSS style rules to be included in the targeted component's local style.
```
Comment (Haprog) [inline README.md]: > accepts a an object

Remove extra "a"

File: register-styles.html [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+<script>
+(function() {
+
+  /**
+   * @namespace Vaadin
+   */
+  window.Vaadin = window.Vaadin || {};
+
+  let moduleIdIndex = 0;
+  // Map of
```
Comment (Haprog) [inline register-styles.html]: Map of what?

File: README.md (hunk 2 above)
Comment (tomivirkki) [reply README.md]: Done.

File: register-styles.html (hunk 3 above)
Comment (tomivirkki) [reply register-styles.html]: Done.

File: register-styles.html [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        ${moduleIncludes.map(include => `<style include=${include}></style>`)}
+      </template>
+    `;
+
+    themeModuleElement.register(themeId);
+  };
+
+  /* Template literal tags derived from LitElement */
+
+  const constructionToken = Symbol();
```
Comment (web-padawan) [inline register-styles.html]: Note, `Symbol` is ES6 feature which needs a separate polyfill. That polyfill is included to `webcomponentsjs` 2.x (Polymer 3 compatible version) but not to 1.x (Polymer 2 & HTML Imports).

We probably need to change this to get Polymer 2 build passing in IE11.

Comment (tomivirkki) [reply register-styles.html]: Right, changed to a plain object

File: README.md [hunk 5]
```diff
@@ -158,7 +158,61 @@ The `id` attribute of the theme module should be unique. You can also re-use an
...
+`, Vaadin.unsafeCSS(trustedCSSValue)]);
+```
+
+> Note: If you need to include styles defined as Polymer style modules with id, you can still pass the module ids as `include` array of the third `options` parameter to the function. Consider this API deprecated though.
+
+```js
+Vaadin.registerStyles('my-element', Vaadin.css`
+  /* Styles which will be included in my-element local scope */
+`, {include: ['my-style-module']});
+```
```
Comment (jouni) [inline README.md]: I would try to make the deprecation notice a little more prominent (and remove the “though”). Perhaps as a code comment or as a prominent “info box/callout”.

Comment (tomivirkki) [reply README.md]: Done.

**Comment by platosha** [comment]
Looks much better now, thanks!

File: register-styles.html (hunk 4 above)
Comment (Haprog) [reply register-styles.html]: @tomivirkki Is this targeting Vaadin 14 or 15? The issue mainly talks about Polymer 3.

If we need this for V14 with IE11 support, then I guess we need to change this part, but if this is targeting only V15+ then I think using Symbol as is is fine?

Comment (tomivirkki) [reply register-styles.html]: This is for V14

**Pull request merged**